kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/GCSToBQWriter.java

```diff
@@ -0,0 +1,178 @@
+   */
+  public GCSToBQWriter(Storage storage,
+                       BigQuery bigQuery,
+                       int retries,
+                       long retryWaitMs) {
+    this.storage = storage;
+    this.bigQuery = bigQuery;
+
+    this.retries = retries;
+    this.retryWaitMs = retryWaitMs;
+  }
+
+  public void writeRows(List<RowToInsert> rows,
+                        TableId tableId,
+                        String bucketName,
+                        String blobName)
+      throws InterruptedException {
+
+    // Get Source URI
+    BlobId blobId = BlobId.of(bucketName, blobName);
+    BlobInfo blobInfo = BlobInfo.newBuilder(blobId).setContentType("text/json").build();
+    String sourceUri = String.format("gs://%s/%s", bucketName, blobName);
+
+    // Check if the table specified exists
+    // This error shouldn't be thrown. All tables should be created by the connector at startup
+    if (bigQuery.getTable(tableId) == null) {
+      throw new ConnectException(
+          String.format("Table with TableId %s does not exist.", tableId.getTable()));
+    }
+
+    // Create a job configuration
+    this.loadJobConfiguration =
+        LoadJobConfiguration.builder(tableId, sourceUri)
+            .setFormatOptions(FormatOptions.json())
+            .setCreateDisposition(JobInfo.CreateDisposition.CREATE_IF_NEEDED)
+            .build();
+
+    int retryCount = 0;
+    boolean exceptionsOccurred;
+    do {
+      if (retryCount > 0) {
+        waitRandomTime();
+      }
+      exceptionsOccurred = false;
+      // Perform GCS Upload and BQ Transfer
+      try {
+        uploadRowsToGcs(rows, blobInfo);
+        triggerBigQueryLoadJob();
+      } catch (ConnectException ce) {
+        exceptionsOccurred = true;
+        logger.warn("Exceptions occurred for table {}, attempting retry", tableId.getTable());
+      }
+      retryCount++;
+    } while (exceptionsOccurred && (retryCount < retries));
+
+    logger.info("Batch loaded {} rows", rows.size());
+  }
+
+  /**
+   * Triggers a load job to transfer JSON records from a GCS blob to a BigQuery table.
+   */
+  private void triggerBigQueryLoadJob() {
+    bigQuery.create(JobInfo.of(loadJobConfiguration));
+  }
+
+  /**
+   * Creates a JSON string containing all records and uploads it as a blob to GCS
+   * @return The blob uploaded to GCS
+   */
+  private Blob uploadRowsToGcs(List<RowToInsert> rows, BlobInfo blobInfo) {
+    try {
+      Blob resultBlob = uploadBlobToGcs(
+          new ByteArrayInputStream(toJson(rows).getBytes("UTF-8")), blobInfo);
+      return resultBlob;
+    } catch (UnsupportedEncodingException uee) {
+      throw new GCSConnectException("Failed to upload blob to GCS", uee);
+    }
+  }
+
+  private Blob uploadBlobToGcs(InputStream blobContent, BlobInfo blobInfo) {
+    // todo look into creating from a string because this is depreciated - input stream cannot retry
+    // todo consider if it would be worth it to switch to a resumable method of uploading
```

Comment:
Can you elaborate on this? I'm fine with committing this as is, but we should probably open an issue with more information this if so so it's easier to track down and change later if we decide it's needed.